### PR TITLE
Propogate exceptions

### PR DESCRIPTION
**Pull request by jamescway**
- propagate auth exceptions to output
- add some exception text to fake cassandra

@oscil8 @trane @rkuris @mbbx6spp

We can now see Authentication exceptions like this:
screen shot 2015-03-25 at 9 00 56 am

### CODE REVIEW

File: zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala [hunk 1]
```diff
@@ -141,7 +141,11 @@ private[cassie] class ClusterClientProvider(
         // call login
         val loginMap: java.util.Map[String, String] = HashMap("username" -> username, "password" -> password)
         val authRequest = new AuthenticationRequest(loginMap)
-        client.login(authRequest)
+        val future = client.login(authRequest)
+        future onFailure { exception =>
+          throw exception
+        }
```
Comment (jamescway) [inline zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala]: unsure how to test this block, as the exceptions get buried inside of futures inside of futures.  I think [ScalaFutures](http://doc.scalatest.org/2.0/index.html#org.scalatest.concurrent.Futures) allow for testing of 1 level of futures but its not enough.  Not to mention these are twitter futures, not scala.  Found a [posting](http://cjwebb.github.io/blog/2015/02/02/scalatest-futures/) about how to convert the futures but I didn't get this to work.
I manually verified that the Auth exceptions which we're trying to catch are shown in the output.

File: zipkin-cassandra/src/schema/cassandra-schema-cql3.txt [hunk 2]
```diff
@@ -6,6 +6,7 @@ CREATE TABLE "Zipkin"."ServiceSpanNameIndex" (
     value blob,
     PRIMARY KEY (key, column1)
 ) WITH CLUSTERING ORDER BY (column1 ASC)
+    AND COMPACT STORAGE
```
Comment (Oscil8) [inline zipkin-cassandra/src/schema/cassandra-schema-cql3.txt]: ah ok ..  had to go read https://www.datastax.com/documentation/cql/3.0/cql/cql_reference/create_table_r.html?scroll=reference_ds_v3f_vfk_xj__using-compact-storage

might be clearer to split this into a separate commit, and describe the issue (thrift API requires this legacy storage method)

File: zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala (hunk 1 above)
Comment (trane) [inline zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala]: what is the thing you want to test? remember, your types will protect you from a lot of things that would normally need to be unit tested in ruby.

File: zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala [hunk 3]
```diff
@@ -141,7 +141,11 @@ private[cassie] class ClusterClientProvider(
         // call login
         val loginMap: java.util.Map[String, String] = HashMap("username" -> username, "password" -> password)
         val authRequest = new AuthenticationRequest(loginMap)
-        client.login(authRequest)
+        val future = client.login(authRequest)
+        future onFailure { exception =>
```
Comment (trane) [inline zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala]: Why are you using side-effects here? Future is already rich enough to not require you to throw exceptions. What are you trying to accomplish with throwing an exception here?

File: zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala (hunk 1 above)
Comment (trane) [inline zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala]: so, i can see where you might want to throw an exception here (if the call to `super.prepareConnFactory(keyspacedFactory)` doesn't raise an exception when `login` or `set_keyspace` return a `Future.exception`. I would make this a bit more explicit, however, in that case:

``` scala
  class CassandraThriftFramedCodec(protocolFactory: TProtocolFactory, config: ClientCodecConfig)
      extends ThriftClientFramedCodec(protocolFactory: TProtocolFactory, config: ClientCodecConfig, None, false) {
    override def prepareConnFactory(factory: ServiceFactory[ThriftClientRequest, Array[Byte]]) = {
      val keyspacedSetFactory: ServiceFactory[ThriftClientRequest, Array[Byte]] = factory flatMap { service =>
        val client = new ServiceToClient(service, new TBinaryProtocol.Factory())
        val login = client.login(new AuthenticationRequest(Map("username" -> username, "password" -> password)))
        val keyspace = client.set_keyspace(keyspace)
        for {
          _ <- login.onFailure(throw _)
          _ <- keyspace.onFailure(throw _)
        } yield service
      }
      // set up tracing
      super.prepareConnFactory(keyspacedSetFactory)
    }
  }
```

Since both `.login` and `.set_keyspace` are of type `Future[Unit]`, so we just bind them to `_` and then make sure that we are explicit in the `.onFailure` calls so that the shape of the code shows you what's going to happen, otherwise we will `yield service` which will return a `Future[Service[...]]`.
I wouldn't be surprised if others disagree (looking at @mbbx6spp) or have a better way :)

File: zipkin-cassandra/src/schema/cassandra-schema-cql3.txt (hunk 2 above)
Comment (jamescway) [inline zipkin-cassandra/src/schema/cassandra-schema-cql3.txt]: Sounds good, I'll split it out

File: zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala (hunk 1 above)
Comment (jamescway) [inline zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala]: I want the exception viewable in the output (which gets turned into the log).  The thinking was that throwing the exception would make it viewable.  Otherwise what's returned is just a future and the exception is hidden.  

The for comprehension looks like a good refactor, but it doesn't seem to make it more testable because this `CassandraThriftFramedCodec` gets buried in multiple levels of futures. 

What I wanted to do initially was to use scalatest to assert the right exception is raised using intercept.  But because of the multiple levels of futures the exception doesn't get bubbled up.

``` scala
intercept[SomeException] {
  ...code...
} 
```

So the question is: [Whats a good way to test this method?]
- one idea is to just make a unit test and see if the exception gets thrown but that doesn't seem to mesh with the rest of the testing strategy which seems more integration-ish

**Comment by mbbx6spp** [comment]
Looks just like we discussed offline to get the desired effect of error logging which was the intended effect. My only other suggestion now is to validate the right number and/or correct error log entry was logged for the test. I am struggling to find where to swap out `com.twitter.logging.Logger` implementations with `TwitterServer` as I am unfamiliar with Twitter APIs, but I can revisit tomorrow perhaps? :) Cheers.

**Comment by jamescway** [comment]
Added a test that intercepts the exception.  Swapped to unexpected exception and test fails.  The only real change is the use of the for-comprehension.

File: zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala [hunk 4]
```diff
@@ -141,8 +144,12 @@ private[cassie] class ClusterClientProvider(
         // call login
         val loginMap: java.util.Map[String, String] = HashMap("username" -> username, "password" -> password)
         val authRequest = new AuthenticationRequest(loginMap)
-        client.login(authRequest)
-        client.set_keyspace(keyspace) map { _ => service }
+        val login = client.login(authRequest)
+        val ks = client.set_keyspace(keyspace)
+        for{
```
Comment (trane) [inline zipkin-cassandra/src/main/scala/com/twitter/cassie/connection/ClusterClientProvider.scala]: ws

**Comment by trane** [comment]
lgtm!
